File: snakebird.go
package main

import (
	"fmt"
)

type point struct {
	x int
	y int
}

type pointSet map[point]bool

func (this *point) eq(that *point) bool {
	return (this == that) || (this.x == that.x && this.y == that.y)
}

type Snakebird []point

type GameState struct {
	snakebird Snakebird
	obstacles pointSet
	fruits    pointSet
	goal      point
}

func (s *Snakebird) side(dx, dy int) pointSet {
	sbMap := make(pointSet)

	for _, sp := range *s {
		sbMap[sp] = true
	}

	result := make(pointSet)

	for _, sp := range *s {
		p := point{sp.x + dx, sp.y + dy}
		if !sbMap[p] {
			result[p] = true
		}
	}

	return result
}

func (s *Snakebird) bottom() pointSet {
	return s.side(0, -1)
}

func (gs *GameState) moveInDir(dx, dy int) {
	for gs.allEmpty(gs.snakebird.side(dx, dy)) {
		newSb := make(Snakebird, len(gs.snakebird))
		copy(newSb, gs.snakebird)
		for i, s := range newSb {
			newSb[i] = point{s.x + dx, s.y + dy}
		}
		gs.snakebird = newSb
	}
}

func (gs *GameState) fall() {
	gs.moveInDir(0, -1)
}

func (gs *GameState) isEmpty(p point) bool {
	if gs.obstacles[p] || p.y < 0 {
		return false
	}
	for _, sp := range gs.snakebird {
		if p == sp {
			return false
		}
	}
	return true
}

func (gs *GameState) allEmpty(ps pointSet) bool {
	for p, _ := range ps {
		if !gs.isEmpty(p) {
			return false
		}
	}
	return true
}

func (this *GameState) Copy() *GameState {
	myFruits := make(pointSet)
	myObstacles := make(pointSet)

	for p, _ := range this.fruits {
		myFruits[p] = true
	}

	for p, _ := range this.obstacles {
		myObstacles[p] = true
	}

	return &GameState{
		goal:      this.goal,
		snakebird: this.snakebird,
		fruits:    myFruits,
		obstacles: myObstacles,
	}
}

func (gs *GameState) moveTo(p point, states *[]GameState) {
	if !gs.isEmpty(p) {
		return
	}

	newGs := *gs.Copy()
	newGs.snakebird = append(newGs.snakebird, p)

	if gs.fruits[p] {
		delete(newGs.fruits, p)
	} else {
		newGs.snakebird = newGs.snakebird[1:]
	}

	newGs.fall()

	*states = append(*states, newGs)
}

func (gs *GameState) move() []GameState {
	current := gs.snakebird[len(gs.snakebird)-1]
	newStates := make([]GameState, 0)
	gs.moveTo(point{current.x + 1, current.y}, &newStates)
	gs.moveTo(point{current.x, current.y + 1}, &newStates)
	gs.moveTo(point{current.x - 1, current.y}, &newStates)
	gs.moveTo(point{current.x, current.y - 1}, &newStates)
	return newStates
}

func NewGameState() GameState {
	return GameState{
		fruits:    make(pointSet),
		obstacles: make(pointSet),
	}
}

func main() {
	gs := NewGameState()
	gs.goal = point{1, 2}
	gs.snakebird = []point{{-1, 0}, {0, 0}}
	gs.fruits[point{1, 0}] = true
	a := gs.move()
	fmt.Println(gs)
	fmt.Println(a)
}

File: snakebird.hs
import qualified Data.Map as M
import qualified Data.Set as S
import Control.Monad

type MyInt = Int
data Point = Point MyInt MyInt deriving (Show, Eq, Ord)

down = Point 0 (-1)
up = Point 0 1
left = Point (-1) 0
right = Point 1 0

instance Num Point where
    (Point x y) + (Point x' y') = Point (x+x') (y+y')
    _ * _ = undefined
    abs = undefined
    fromInteger n = Point (fromIntegral n) 0
    negate (Point x y) = Point (-x) (-y)
    signum = undefined

type PointSet = S.Set Point

type Snakebird = [Point]

data GameState = GameState {
    snakebird :: Snakebird,
    obstacles :: PointSet,
    fruits :: PointSet,
    goal :: Point
} deriving (Show, Eq, Ord)

side :: Point -> Snakebird -> [Point]
side p s = filter (`elem` s) $ map (+p) s

bottom = side down

isEmpty :: GameState -> Point -> Bool
isEmpty gs p@(Point x y) = and [y >= 0, not $ S.member p $ obstacles gs, not $ p `elem` (snakebird gs)]

allEmpty :: GameState -> [Point] -> Bool
allEmpty gs = null . filter (not . isEmpty gs)

moveSnakebird :: Point -> GameState -> GameState
moveSnakebird p gs = if allEmpty gs (side p (snakebird gs))
    then moveSnakebird p $ gs { snakebird = map (+p) (snakebird gs) }
    else gs

fall = moveSnakebird down

moveHead :: GameState -> Point -> [GameState]
moveHead gs p = do
    guard $ isEmpty gs p
    let hasFruit = S.member p (fruits gs)
    let newFruits = (if not hasFruit then id else S.delete p) (fruits gs)
    let newSnakebird = (if hasFruit then id else init) (p : snakebird gs)
    return $ gs { fruits = newFruits, snakebird = newSnakebird }

move :: GameState -> [GameState]
move gs = do
    dir <- [right, up, left, down]
    moveHead gs dir

main = do
    print gs
    mapM_ print (move gs)
    where
        gs = GameState [Point 0 0, Point (-1) 0] S.empty (S.singleton (Point 1 0)) (Point 1 2)
